File: python/quantlib/trinomial.py
"""
Trinomial tree of Kamrad-Ritchken for European and American options.

Phase 3 Block 5. Implements the lambda-parametrised trinomial tree:

    x -> x + dx  with prob p_u = 1/(2 lambda) + nu sqrt(dt) / (2 sigma sqrt(lambda))
    x -> x       with prob p_m = 1 - 1/lambda
    x -> x - dx  with prob p_d = 1/(2 lambda) - nu sqrt(dt) / (2 sigma sqrt(lambda))

with dx = sigma sqrt(lambda dt) and nu = r - sigma^2/2.

The default lambda = 3 corresponds to (p_u, p_m, p_d) ~ (1/6, 2/3, 1/6),
the weights of Simpson's rule, which match the third moment of the
GBM step. For lambda = 1, p_m = 0 and the tree degenerates to a
binomial-equivalent.

The trinomial backward induction is structurally equivalent to the
FTCS finite-difference scheme of Block 1 with alpha = 1/(2 lambda).
This is a structural insight, not a coincidence: see Block 5 writeup
for the explicit identification of coefficients.

References
----------
Kamrad and Ritchken (1991), Management Science 37(12), 1640-1652.
"""

import numpy as np


def _validate_inputs(S, K, r, sigma, T, n_steps, lambda_param):
    """Validate inputs common to all trinomial pricers.

    Raises ValueError on any invalid combination.
    """
    if S <= 0:
        raise ValueError(f"S must be positive, got {S}")
    if K <= 0:
        raise ValueError(f"K must be positive, got {K}")
    if sigma <= 0:
        raise ValueError(f"sigma must be positive, got {sigma}")
    if T <= 0:
        raise ValueError(f"T must be positive, got {T}")
    if n_steps < 1:
        raise ValueError(f"n_steps must be >= 1, got {n_steps}")
    if lambda_param < 1.0:
        raise ValueError(
            f"lambda_param must be >= 1 (else p_m < 0); got {lambda_param}"
        )


def _trinomial_probabilities(r, sigma, dt, lambda_param):
    """Compute the KR probabilities p_u, p_m, p_d.

    Returns
    -------
    (p_u, p_m, p_d) : tuple of float
        Each in [0, 1] under the validation conditions above. Their sum
        is 1 by construction.

    Notes
    -----
    For project parameters (|nu| sqrt(dt) << sigma sqrt(lambda)) the
    drift correction is small. We do not check positivity of p_d at
    runtime: if it failed, it would only be because the user picked
    pathological parameters (extreme dt, near-zero sigma, etc.) and
    the diagnostic would surface as an immediately visible bug.
    """
    nu = r - 0.5 * sigma * sigma
    drift_term = nu * np.sqrt(dt) / (2.0 * sigma * np.sqrt(lambda_param))
    p_u = 0.5 / lambda_param + drift_term
    p_m = 1.0 - 1.0 / lambda_param
    p_d = 0.5 / lambda_param - drift_term
    return p_u, p_m, p_d


def _trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param,
                        payoff_type):
    """Run the backward induction for any trinomial pricer.

    Parameters
    ----------
    payoff_type : str
        One of 'european_call', 'european_put', 'american_put'.

    Returns
    -------
    float
        Price at the root of the tree.

    Notes
    -----
    Uses an in-place array of length 2 * n_steps + 1 for memory
    efficiency. Cost: O(n_steps^2) flops, O(n_steps) memory.
    """
    dt = T / n_steps
    dx = sigma * np.sqrt(lambda_param * dt)
    p_u, p_m, p_d = _trinomial_probabilities(r, sigma, dt, lambda_param)
    disc = np.exp(-r * dt)

    # Terminal payoffs at j = -n_steps, ..., +n_steps.
    n_nodes = 2 * n_steps + 1
    j_grid = np.arange(-n_steps, n_steps + 1, dtype=np.float64)
    S_terminal = S * np.exp(j_grid * dx)
    if payoff_type == 'european_call':
        V = np.maximum(S_terminal - K, 0.0)
    elif payoff_type in ('european_put', 'american_put'):
        V = np.maximum(K - S_terminal, 0.0)
    else:
        raise ValueError(f"unknown payoff_type: {payoff_type!r}")

    # Backward induction. At step n we have 2*(N - n) + 1 active nodes;
    # we shrink the array by 2 per step (one off each end).
    is_american = (payoff_type == 'american_put')
    # We reuse the same buffer, shrinking by reading from a sliced view.
    for n in range(n_steps - 1, -1, -1):
        # New centre is at the same position; new j range is -n..+n.
        # Active nodes at step n+1 are V[start:end] where start = n_steps - (n+1)
        # and end = n_steps + (n+1) + 1.
        # New nodes at step n correspond to indices V[start+1:end-1].
        # The recurrence reads V[i-1], V[i], V[i+1] (at step n+1) to
        # produce V_new[i] (at step n).
        # We compute new values into a fresh slice and write them back.
        start = n_steps - (n + 1)
        end   = n_steps + (n + 1) + 1
        # Continuation values for the next layer (length 2n+1).
        cont = disc * (
            p_u * V[start + 2 : end]
          + p_m * V[start + 1 : end - 1]
          + p_d * V[start     : end - 2]
        )
        if is_american:
            # Spot prices at this layer's nodes: j = -n .. +n.
            j_n = np.arange(-n, n + 1, dtype=np.float64)
            S_n = S * np.exp(j_n * dx)
            exer = np.maximum(K - S_n, 0.0)
            V_new = np.maximum(cont, exer)
        else:
            V_new = cont
        # Write back into the corresponding slice of V.
        V[start + 1 : end - 1] = V_new

    # The root is at index n_steps (the centre of the original array).
    return float(V[n_steps])


# ---------------------------------------------------------------------------
# High-level pricers
# ---------------------------------------------------------------------------
def trinomial_european_call(S, K, r, sigma, T, *, n_steps,
                            lambda_param=3.0):
    """Price a European call by the Kamrad-Ritchken trinomial tree.

    Parameters
    ----------
    S, K, r, sigma, T : float
        Standard Black-Scholes parameters.
    n_steps : int (keyword-only)
        Number of time steps in the tree.
    lambda_param : float, default 3.0
        Stretching parameter; must be >= 1. Default 3 is the standard
        Kamrad-Ritchken choice. Set to 1 to recover binomial-equivalent
        behaviour (and exact equivalence with FTCS).

    Returns
    -------
    float
        Trinomial approximation to the Black-Scholes call price.
    """
    _validate_inputs(S, K, r, sigma, T, n_steps, lambda_param)
    return _trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param,
                                'european_call')


def trinomial_european_put(S, K, r, sigma, T, *, n_steps,
                           lambda_param=3.0):
    """Price a European put by the Kamrad-Ritchken trinomial tree.
    See trinomial_european_call.
    """
    _validate_inputs(S, K, r, sigma, T, n_steps, lambda_param)
    return _trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param,
                                'european_put')


def trinomial_american_put(S, K, r, sigma, T, *, n_steps,
                           lambda_param=3.0):
    """Price an American put by the Kamrad-Ritchken trinomial tree.

    The early-exercise check is applied at every node at every step:
    V_node = max(intrinsic value, discounted continuation value).

    Note: there is no trinomial_american_call function because, for a
    non-dividend-paying stock, the American call equals the European
    call (Merton 1973).
    """
    _validate_inputs(S, K, r, sigma, T, n_steps, lambda_param)
    return _trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param,
                                'american_put')


# ---------------------------------------------------------------------------
# Smoke test
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    # Hull example 15.6: S=42, K=40, r=0.10, sigma=0.20, T=0.5
    # BS European call: 4.7594, put: 0.8086
    S, K, r, sigma, T = 42.0, 40.0, 0.10, 0.20, 0.5
    print(f"Hull 15.6: S={S}, K={K}, r={r}, sigma={sigma}, T={T}")
    print(f"  BS reference: call=4.7594, put=0.8086\n")

    print("Trinomial KR (lambda=3, default):")
    print(f"  European call (n=2000): "
          f"{trinomial_european_call(S, K, r, sigma, T, n_steps=2000):.6f}")
    print(f"  European put  (n=2000): "
          f"{trinomial_european_put (S, K, r, sigma, T, n_steps=2000):.6f}")
    print(f"  American put  (n=2000): "
          f"{trinomial_american_put (S, K, r, sigma, T, n_steps=2000):.6f}")
    print()
    print("Trinomial KR (lambda=1, binomial-equivalent):")
    print(f"  European call (n=2000): "
          f"{trinomial_european_call(S, K, r, sigma, T, n_steps=2000, lambda_param=1.0):.6f}")

File: cpp/src/trinomial.cpp
// trinomial.cpp - implementation of the Kamrad-Ritchken trinomial tree.

#include "trinomial.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <string>
#include <vector>

namespace quant {
namespace pde {

namespace {

// ---------------------------------------------------------------------
// Common input validation.
// ---------------------------------------------------------------------
void validate_inputs(double S, double K, double r, double sigma, double T,
                     int n_steps, double lambda_param) {
    (void)r;  // r can be any sign in principle; no constraint
    if (S <= 0.0) {
        throw std::invalid_argument(
            "trinomial: S must be positive, got " + std::to_string(S));
    }
    if (K <= 0.0) {
        throw std::invalid_argument(
            "trinomial: K must be positive, got " + std::to_string(K));
    }
    if (sigma <= 0.0) {
        throw std::invalid_argument(
            "trinomial: sigma must be positive, got " + std::to_string(sigma));
    }
    if (T <= 0.0) {
        throw std::invalid_argument(
            "trinomial: T must be positive, got " + std::to_string(T));
    }
    if (n_steps < 1) {
        throw std::invalid_argument(
            "trinomial: n_steps must be >= 1, got " + std::to_string(n_steps));
    }
    if (lambda_param < 1.0) {
        throw std::invalid_argument(
            "trinomial: lambda_param must be >= 1 (else p_m < 0), got " +
            std::to_string(lambda_param));
    }
}

// ---------------------------------------------------------------------
// Backward induction core. payoff_type:
//   0 = european_call, 1 = european_put, 2 = american_put.
//
// The recurrence V_new[i] = f(V_old[i-1], V_old[i], V_old[i+1]) requires
// a separate buffer for the new values: an in-place sweep in either
// direction would corrupt at least one of the three reads. We use a
// pre-allocated scratch buffer of the same size as V to keep memory
// allocation out of the hot loop.
// ---------------------------------------------------------------------
double trinomial_backward(double S, double K, double r, double sigma,
                          double T, int n_steps, double lambda_param,
                          int payoff_type) {
    const double dt = T / static_cast<double>(n_steps);
    const double dx = sigma * std::sqrt(lambda_param * dt);
    const double nu = r - 0.5 * sigma * sigma;

    const double drift_term =
        nu * std::sqrt(dt) / (2.0 * sigma * std::sqrt(lambda_param));
    const double p_u = 0.5 / lambda_param + drift_term;
    const double p_m = 1.0 - 1.0 / lambda_param;
    const double p_d = 0.5 / lambda_param - drift_term;
    const double disc = std::exp(-r * dt);

    // Terminal payoffs: 2*n_steps + 1 nodes at j = -n_steps, ..., +n_steps.
    const std::size_t n_nodes = static_cast<std::size_t>(2 * n_steps + 1);
    std::vector<double> V(n_nodes);
    std::vector<double> V_scratch(n_nodes);

    for (std::size_t idx = 0; idx < n_nodes; ++idx) {
        const int j = static_cast<int>(idx) - n_steps;
        const double S_j = S * std::exp(static_cast<double>(j) * dx);
        switch (payoff_type) {
            case 0:  // european call
                V[idx] = std::max(S_j - K, 0.0);
                break;
            case 1:  // european put
            case 2:  // american put (same terminal payoff)
                V[idx] = std::max(K - S_j, 0.0);
                break;
            default:
                throw std::invalid_argument(
                    "trinomial_backward: unknown payoff_type");
        }
    }

    const bool is_american = (payoff_type == 2);

    // Backward induction. At step n the active nodes occupy V[start..end).
    for (int n = n_steps - 1; n >= 0; --n) {
        const std::size_t start = static_cast<std::size_t>(n_steps - (n + 1));
        const std::size_t end   = static_cast<std::size_t>(n_steps + (n + 1) + 1);

        // Each new node at step n (j = -n .. +n) is at index in (start, end-1).
        for (std::size_t i = start + 1; i + 1 < end; ++i) {
            const double cont =
                disc * (p_u * V[i + 1] + p_m * V[i] + p_d * V[i - 1]);

            if (is_american) {
                const int j_node = static_cast<int>(i) - n_steps;
                const double S_node =
                    S * std::exp(static_cast<double>(j_node) * dx);
                const double exer = std::max(K - S_node, 0.0);
                V_scratch[i] = std::max(cont, exer);
            } else {
                V_scratch[i] = cont;
            }
        }

        // Copy the new layer back into V. Only the active range is updated.
        for (std::size_t i = start + 1; i + 1 < end; ++i) {
            V[i] = V_scratch[i];
        }
    }

    // The root is at the centre index n_steps.
    return V[static_cast<std::size_t>(n_steps)];
}

}  // namespace

// ---------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------
double trinomial_european_call(double S, double K, double r, double sigma,
                                double T, int n_steps, double lambda_param) {
    validate_inputs(S, K, r, sigma, T, n_steps, lambda_param);
    return trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param, 0);
}

double trinomial_european_put(double S, double K, double r, double sigma,
                                double T, int n_steps, double lambda_param) {
    validate_inputs(S, K, r, sigma, T, n_steps, lambda_param);
    return trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param, 1);
}

double trinomial_american_put(double S, double K, double r, double sigma,
                                double T, int n_steps, double lambda_param) {
    validate_inputs(S, K, r, sigma, T, n_steps, lambda_param);
    return trinomial_backward(S, K, r, sigma, T, n_steps, lambda_param, 2);
}

}  // namespace pde
}  // namespace quant
